engineering/architecture/front-end-standards.md

```diff
@@ -77,4 +77,219 @@ dateCreated: 2022-01-14T19:02:19.924Z
+
+    ```svelte
+    <Checkbox value={null} />
+    ```
+
+Considerations:
+
+- In some cases you may need to coalesce a `null` value to `undefined`. For example:
+
+    ```ts
+    function firstCapitalLetter(s: string): string | undefined {
+    return s.match(/[A-Z]/)?.[0] ?? undefined
+    }
+    ```
+
+Additional context:
+
+- [discussion](https://github.com/centerofci/mathesar/discussions/825)
+
+## Minimize Svelte store instances
+
+- ✅ Good because only one `cost` store is created.
+
+    ```ts
+    function getCost(toppings: Readable<string[]>) {
+      return derived(this.toppings, t => 10 + t.length * 2);
+    }
+
+    class PizzaOrder {
+      toppings: string[];
+      cost: Readable<number>;
+
+      constructor() {
+        this.toppings = [];
+        this.cost = getCost(this.toppings);
+      }
+    }
+    ```
+
+- ❌ Bad because separate calls to `cost` will create separate stores which may lead to more subscribe and unsubscribe events in some cases, risking performance problems.
+
+    ```ts
+    class PizzaOrder {
+      toppings: string[];
+
+      constructor() {
+        this.toppings = [];
+      }
+
+      get cost(): Readable<boolean> {
+        return derived(this.toppings, t => 10 + t.length * 2);
+      }
+    }
+    ```
+
+Additional context:
+
+- [Discussion](https://github.com/centerofci/mathesar/pull/776#issuecomment-963831424)
+
+## Build components for valid DOM nesting
+
+When working inside a component that _might_ be placed where [phrasing content](https://developer.mozilla.org/en-US/docs/Web/Guide/HTML/Content_categories#phrasing_content) is required, be sure to only use phrasing content elements (like `span`) instead of _not_-phrasing content elements (like `div`).
+
+-  ✅ Good
+
+    `FancyCheckbox.svelte`
+
+    ```ts
+    <span class="make-it-fancy">
+      <input type="checkbox" />
+    </span>
+    ```
+
+- ❌ Bad because it uses `div` instead of `span`
+
+    `FancyCheckbox.svelte`
+
+    ```ts
+    <div class="make-it-fancy">
+      <input type="checkbox" />
+    </div>
+    ```
+
+Rationale:
+
+- For example, let's build on the "Bad" example above and write the following Svelte code:
+
+    ```svelte
+    <label>
+      <FancyCheckbox />
+      Foo
+    </label>
+    ```
+
+    That Svelte code will render:
+
+    ```html
+    <label>
+      <div class="make-it-fancy">
+        <input type="checkbox" />
+      </div>
+      Foo
+    </label>
+    ```
+
+    That markup has invalid DOM nesting because a `label` element can only contain phrasing content -- but a `div` is _not_ phrasing content.
+
+Tips:
+
+- You can make a `span` _act_ like a `div` by setting `display: block` if needed.
+
+Notes:
+
+- Not all of our components adhere to this guideline yet.
+
+
+## `type` vs `interface`
+
+- Prefer `interface` when possible.
+- Use `type` when necessary.
+
```

Comment:
@pavish this is some content that we haven't discussed before, but from reading your code I've deduced that we're on the same page about this (which is also the direction that I think the large TS community is going).